### ISSUE TITLE
.cpu() hangs after gather_for_metrics

### ISSUE THREAD

**Issue by ahmadshapiro** [huggingface/accelerate#1369]
### System Info

```Shell
This my evaluation loop which is edited from the official deepspeed example.
I'm training on 4 nodes of 4 GPUs each on slurm. 


def evaluate(cfg, model, tokenizer, eval_dataloader, accelerator, eval_dataset, logger):
    progress_bar = tqdm(range(len(eval_dataloader)//cfg.training.per_device_eval_batch_size), disable=not accelerator.is_main_process)
    torch.cuda.empty_cache()
    model.eval()
    losses = []
    logger.info("Starting Evaluation")
    samples_seen = 0
    for step, batch in enumerate(eval_dataloader):
        with torch.no_grad():
            outputs = model(**batch)
        loss = outputs.loss
        labels = torch.where(batch['labels'] != -100, batch['labels'], 0)
        preds = torch.where(batch['labels'] != -100, torch.argmax(outputs['logits'], dim=-1), 0)
        
        logger.info("Gathering Predictions and Labels .. ")
        preds, labels = accelerator.gather_for_metrics(preds), accelerator.gather_for_metrics(labels)
        logger.info(f"Gathered Predictions {type(preds)} on {preds.device} and Labels {type(labels)} on {labels.device}")
        
        logger.info("Gathering Losses .. ")
        losses.append(accelerator.gather_for_metrics(loss.repeat(cfg.training.per_device_eval_batch_size)))
        

        logger.info("Moving Predictions and Labels .. ")
        preds, labels = preds.cpu().numpy(), labels.cpu().numpy()
        
        logger.info("Decoding gathered predictions and labels")
        decoded_labels = tokenizer.batch_decode(labels, skip_special_tokens=True)
        decoded_preds = tokenizer.batch_decode(preds, skip_special_tokens=True)
        logger.info("Done decoding")
        rouge.add_batch(predictions = decoded_preds, references=decoded_labels)
        logger.info(f"Added a batch to metric of {len(decoded_labels)} labels and {len(decoded_predictions)} predicitions")
        del preds, labels
        torch.cuda.empty_cache()
            
        progress_bar.update(1)
        
    losses = torch.cat(losses)
    try:
        eval_loss = torch.mean(losses)
        perplexity = math.exp(eval_loss)
        rouge_results = rouge.compute()
    except OverflowError:
        perplexity = float("inf")
    return perplexity, eval_loss, rouge_results['rouge1'], rouge_results['rouge2'], rouge_results['rougeL'], rouge_results['rougeLsum']
```
After gathering the predictions and labels, the code hangs when i try to move them to cpu or when i call accelerator.wait_for_everyone().
I thought that some process might haven't gathered yet, so i printed print("Process {accelerator.process_index} gathered"). All of the 16 (4 nodes x 4 GPUs) processes has gathered. but it still hangs at both .wait_for_everyone()``` and ```preds, labels = preds.cpu().numpy(), labels.cpu().numpy()``` i tried a variants of .cpu() like to('cpu') but it also hanfs.
```


### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [X] One of the scripts in the examples/ folder of Accelerate or an officially supported `no_trainer` script in the `examples` folder of the `transformers` repo (such as `run_no_trainer_glue.py`)
- [ ] My own task or dataset (give details below)

### Reproduction

def evaluate(cfg, model, tokenizer, eval_dataloader, accelerator, eval_dataset, logger):
    progress_bar = tqdm(range(len(eval_dataloader)//cfg.training.per_device_eval_batch_size), disable=not accelerator.is_main_process)
    torch.cuda.empty_cache()
    model.eval()
    losses = []
    logger.info("Starting Evaluation")
    samples_seen = 0
    for step, batch in enumerate(eval_dataloader):
        with torch.no_grad():
            outputs = model(**batch)
        loss = outputs.loss
        labels = torch.where(batch['labels'] != -100, batch['labels'], 0)
        preds = torch.where(batch['labels'] != -100, torch.argmax(outputs['logits'], dim=-1), 0)
        
        logger.info("Gathering Predictions and Labels .. ")
        preds, labels = accelerator.gather_for_metrics(preds), accelerator.gather_for_metrics(labels)
        logger.info(f"Gathered Predictions {type(preds)} on {preds.device} and Labels {type(labels)} on {labels.device}")
        
        logger.info("Gathering Losses .. ")
        losses.append(accelerator.gather_for_metrics(loss.repeat(cfg.training.per_device_eval_batch_size)))
        

        logger.info("Moving Predictions and Labels .. ")
        preds, labels = preds.cpu().numpy(), labels.cpu().numpy()
        
        logger.info("Decoding gathered predictions and labels")
        decoded_labels = tokenizer.batch_decode(labels, skip_special_tokens=True)
        decoded_preds = tokenizer.batch_decode(preds, skip_special_tokens=True)
        logger.info("Done decoding")
        rouge.add_batch(predictions = decoded_preds, references=decoded_labels)
        logger.info(f"Added a batch to metric of {len(decoded_labels)} labels and {len(decoded_predictions)} predicitions")
        del preds, labels
        torch.cuda.empty_cache()
            
        progress_bar.update(1)
        
    losses = torch.cat(losses)
    try:
        eval_loss = torch.mean(losses)
        perplexity = math.exp(eval_loss)
        rouge_results = rouge.compute()
    except OverflowError:
        perplexity = float("inf")
    return perplexity, eval_loss, rouge_results['rouge1'], rouge_results['rouge2'], rouge_results['rougeL'], rouge_results['rougeLsum']

### Expected behavior

It should not hang.

**Comment by muellerzr** [comment]
Does it hang on "Gathering losses" or "Moving predictions and labels"

**Comment by ahmadshapiro** [comment]
"Moving predictions and labels"

**Comment by ahmadshapiro** [comment]
Turned out it was my mistake, i didn't know that both preds and labels should of the same size (sequence length) across all GPUs. which is wrong in my case. Solved it via padding
